cfd_protocol/src/lib.rs

```diff
@@ -209,12 +208,157 @@ pub fn spending_tx_sighash(
 }
 
 pub fn finalize_spend_transaction(
-    _tx: Transaction,
-    _commit_descriptor: Descriptor<PublicKey>,
-    (_maker_pk, _maker_sig): (PublicKey, Signature),
-    (_taker_pk, _taker_sig): (PublicKey, Signature),
-) -> Transaction {
-    todo!()
+    tx: Transaction,
+    spent_descriptor: &Descriptor<PublicKey>,
+    (maker_pk, maker_sig): (PublicKey, Signature),
+    (taker_pk, taker_sig): (PublicKey, Signature),
+) -> Result<Transaction> {
+    let satisfier = {
+        let mut satisfier = HashMap::with_capacity(2);
+
+        satisfier.insert(maker_pk, (maker_sig, SigHashType::All));
+        satisfier.insert(taker_pk, (taker_sig, SigHashType::All));
+
+        satisfier
+    };
+
+    let mut input = tx
+        .input
+        .clone()
+        .into_iter()
+        .exactly_one()
+        .expect("all spend transactions to have one input");
+    spent_descriptor.satisfy(&mut input, satisfier)?;
+
+    let mut tx = tx;
+    tx.input = vec![input];
+
+    Ok(tx)
+}
+
+pub fn punish_transaction(
+    commit_descriptor: &Descriptor<PublicKey>,
+    address: &Address,
+    encsig: EcdsaAdaptorSignature,
+    sk: SecretKey,
+    revocation_them_sk: SecretKey,
+    publish_them_pk: PublicKey,
+    revoked_commit_tx: &Transaction,
+) -> Result<Transaction> {
+    // CommitTransaction has only one input
+    let input = revoked_commit_tx.input.clone().into_iter().exactly_one()?;
+
+    // Extract all signatures from witness stack
+    let mut sigs = Vec::new();
+    for witness in input.witness.iter() {
+        let witness = witness.as_slice();
+
+        let res = bitcoin::secp256k1::Signature::from_der(&witness[..witness.len() - 1]);
+        match res {
+            Ok(sig) => sigs.push(sig),
+            Err(_) => {
+                continue;
+            }
+        }
+    }
+
+    if sigs.is_empty() {
+        // No signature found, this should fail
+        unimplemented!()
+    }
+
+    // Attempt to extract y_other from every signature
+    let publish_them_sk = sigs
+        .into_iter()
+        .find_map(|sig| encsig.recover(SECP256K1, &sig, &publish_them_pk.key).ok())
+        .context("Could not recover secret key from revoked transaction")?;
```

Comment:
Wild idea: To condense this down, could extract a custom `Iterator` extension trait that works on `Iterator<Item = Signature>`.